### DISCUSSION TITLE
Problems in setting Multifab with ParallelFor

### DISCUSSION THREAD

**Discussion by solarmaterial** [AMReX-Codes/amrex#3970] [Q&A] [answered]
Hey AMReX team,
  I am trying to set data in the Multifab `rhs` based on the data from the Multifab `phi_old`. When the value of  the `phi(i,j,k,0)` is around 0.001, I'd like to store the index number of the x-direction `i0`, and compute the value of the physical domain.  Then another index number `i1` is calculated. In the physical domain defined by these two index numbers, e.g.` (x1,x0)`, I want to set `rhs_array(i,j,k) = 1.`
   ```
GpuArray<Real,AMREX_SPACEDIM> dx = geom.CellSizeArray();
GpuArray<Real,AMREX_SPACEDIM> prob_lo = geom.ProbLoArray();

 rhs.setVal(0,0);
 for ( MFIter mfi(rhs); mfi.isValid(); ++mfi )
    {
        const Box& bx = mfi.validbox();
        const Array4<Real>& rhs_array = rhs.array(mfi);
        const Array4<Real>& phi = phi_old.array(mfi);

        ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k)
        {
            Real x0,x1;
            int i0, i1;      

            if ( phi(i,j,k,0) - 0.001 < 0. && phi(i+1,j,k,0) - 0.001 > 0.)
            {
                i0 = i;
                x0 = prob_lo[0] + (i0+Real(0.5)) * dx[0];
                x1 = x0 - 2.e-3;
                i1 = (x1 - prob_lo[0]) / dx[0] - Real(0.5);
            }

            if (i1< i  &&  i <= i0)
            {
             rhs_array(i,j,k) = 1.;
            }
                             
	    });
    }
```
However, when the` n_cell ` is not equal to `max_grid_size`, such as when `n_cell = 128, max_grid_size=32`, the output of the code above is not what I expect. This might be the compute domain is cut to 4 parts, I guess? I wonder what I can do to fix this issue.  I'm really looking forward to your reply!

Yours sincerely,
solar

**Comment by WeiqunZhang** [comment]
If the first `if` test is false, `i0` and `i1` are uninitialized. That could be the issue.

(I assume `rhs` has already be initialized. Otherwise if the second `if` test is false, `rhs` in those cells is also uninitialized.)

**Reply by WeiqunZhang** [reply]
By the way, if you compile with warnings on (e.g., `make WARN_ALL=TRUE` with GNU make), the compiler should be able to give warnings about it.

**Reply by solarmaterial** [reply]
Thank you for your replies! Well, I do check the prints of the `i0` and `i1`, I'm pretty sure that they are initialized properly. Here are some details.  
 ```
 rhs.setVal(0,0);
....
 
  if ( phi(i,j,k,0) - 0.003 < 0. && phi(i+1,j,k,0) - 0.003 > 0.)
            {
                x0 = prob_lo[0] + (i+Real(0.5)) * dx[0];
                x1 = x0 - 2.e-3;
                
                int i_mid = std::lround( ( x1 - prob_lo[0] ) / dx[0] - Real(0.5) );

                rhs_array(i,j,k) = i;
                rhs_array(i+1,j,k) = prob_lo[0];
                rhs_array(i+2,j,k) = x0;
                rhs_array(i+3,j,k) = x1;
                rhs_array(i+4,j,k) = Cv;
                rhs_array(i+5,j,k) = i_mid ;
                
                i1 = i;
                i0 = i_mid ;
            }
```
I used the codes above to print the Multifab of the `rhs` to check the values of the `i0` and `i1`. if I set the `max_grid_size` is big enough, like `n_cell = 1200`, `max_grid_size = 100`. The outputs of the `rhs` can clearly show the values of the 6 points (from `rhs_array(i,j,k)` to `rhs_array(i+5,j,k)`)

```
......
    -0.025583333333333333                        0
    -0.025416666666666657                        0
    -0.025249999999999995                      448
    -0.025083333333333332     -0.10000000000000001
    -0.024916666666666656    -0.025249999999999995
    -0.024749999999999994    -0.027249999999999996
    -0.024583333333333332                       80
    -0.024416666666666656                      436
    -0.024249999999999994                        0
    -0.024083333333333332                        0 
......
```
However, when I set `n_cell = 1200` and `max_grid_size = 10`, the outputs become strange, with some values missing. I suspect this issue is related to how the computational domain is divided using `max_grid_size` by `ParallelFor`. The index value of `i` may be changing during the division process.
```
    -0.025749999999999995                        0
    -0.025583333333333333                        0
    -0.025416666666666657                        0
    -0.025249999999999995                      448
    -0.025083333333333332     -0.10000000000000001
    -0.024916666666666656                        0
    -0.024749999999999994                        0
    -0.024583333333333332                        0
```
I'm still looking forward to your reply and would greatly appreciate your assistance!

**Reply by WeiqunZhang** [reply]
You code you post the second time seems very different from the first time. In the first code, i0 and i1 are certainly not always initialized. The second code is incomplete. So I cannot comment on it.

**Reply by solarmaterial** [reply]
I have updated my code.
```
GpuArray<Real,AMREX_SPACEDIM> dx = geom.CellSizeArray();
GpuArray<Real,AMREX_SPACEDIM> prob_lo = geom.ProbLoArray();

rhs.setVal(0,0);
for ( MFIter mfi(rhs); mfi.isValid(); ++mfi )
 {
     const Box& bx = mfi.validbox();
     const Array4<Real>& rhs_array = rhs.array(mfi);
     const Array4<Real>& phi = phi_old.array(mfi);

     ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k)
     {
         Real x0,x1;
         int i0, i1;      

 if ( phi(i,j,k,0) - 0.003 < 0. && phi(i+1,j,k,0) - 0.003 > 0.)
           {
               x0 = prob_lo[0] + (i+Real(0.5)) * dx[0];
               x1 = x0 - 2.e-3;
               
               int i_mid = std::lround( ( x1 - prob_lo[0] ) / dx[0] - Real(0.5) );

               rhs_array(i,j,k) = i;
               rhs_array(i+1,j,k) = prob_lo[0];
               rhs_array(i+2,j,k) = x0;
               rhs_array(i+3,j,k) = x1;
               rhs_array(i+4,j,k) = Cv;
               rhs_array(i+5,j,k) = i_mid ;
               
               i1 = i;
               i0 = i_mid ;
           }
                          
     });
 }
```

**Reply by solarmaterial** [reply]
And if I add another `if ` test, 
```
         if (i1< i  &&  i <= i0)
         {
          rhs_array(i,j,k) = 1.;
         }
```
The output of `rhs `shows that the value 1. is not within the index range of `436 `to `448`.

**Reply by solarmaterial** [reply]
Sorry for the typo. The add `if` test should be
```
         if (i0< i  &&  i <= i1)
         {
          rhs_array(i,j,k) = 1.;
         }
```

**Reply by WeiqunZhang** [reply]
I am sorry. This is very confusing. Maybe you could put your code (at least the function containing part of the code you have shown) on github?

Nevertheless,
```
               rhs_array(i,j,k) = i;
               rhs_array(i+1,j,k) = prob_lo[0];
               rhs_array(i+2,j,k) = x0;
               rhs_array(i+3,j,k) = x1;
               rhs_array(i+4,j,k) = Cv;
               rhs_array(i+5,j,k) = i_mid ;
```
does not seem right. It's inside if (...) test. What if the test is true for both i = 10 and i = 11? What should be value of `rhs_array(11,j,k)` be? Mabye you want to use a six component multifab? The you can have rhs_array(i,j,k,0), rhs_array(i,j,k,1), .... rhs_array(i,j,k,5).

**Reply by solarmaterial** [reply]
Thanks for your replies! The values of the data from `phi_old ` is set to be x-dir dependent（e.g. `phi_old(x,y,z) = f(x)`, and the value is increasing with the growing `x`(e.g. `i`), thus the test is true for only one `i`. I've also uploaded a simple code to github. If I run`./fextract.gnu.ex -y 0 -z 0 rhs`, and run `./fextract.gnu.ex -y 0 -z 1 rhs`, the outputs of these two commands should be the same, but they are not. This issue bothers me a lot. Thank you for your kindly help again!



Appendix:
output of the `./fextract.gnu.ex -y 0 -z 0 rhs`
```
# 1-d slice in x-direction, file: rhs
# time =                    0
#                       x                      666
                 0.015625                        0
                 0.046875                        0
                 0.078125                        0
                 ......                       ......(0)
                 0.453125                        0
                 0.484375                        0
                 0.515625                        0
                 0.546875                      666
                 0.578125                        0
                 0.609375                        0
                 ......                       ......(0)
                 0.921875                        0
                 0.953125                        0
                 0.984375                        0
```
output of the `./fextract.gnu.ex -y 0 -z 1 rhs`
```
# 1-d slice in x-direction, file: rhs
# time =                    0
#                       x                      666
                 0.015625                        0
                 0.046875                        0
                 0.078125                        0
                 ......                       ......(0)
                 0.359375                        0
                 0.390625                        0
                 0.421875                      666
                 0.453125                      666
                 0.484375                      666
                 0.515625                      666
                 0.546875                      666
                 0.578125                        0
                 0.609375                        0
                 ......                       ......(0)
                 0.921875                        0
                 0.953125                        0
                 0.984375                        0
```

**Reply by WeiqunZhang** [reply]
> I've also uploaded a simple code to github.

But could you share a link so that we can better help you?  How many ghost cells does rhs have? From the various pieces you have shown, it probably needs 5. And after you fill rhs, you will need to call `SumBounary` so that the ghost cell values can be added to valid cells in other boxes. The special point might be near the box boundary such that i+1, or i+2, ... or i+5 is in the ghost cell region.

**Reply by solarmaterial** [reply]
The link is here: https://github.com/solarmaterial/amrex
Thanks a lot!

**Reply by WeiqunZhang** [reply]
> If the first if test is false, i0 and i1 are uninitialized. That could be the issue.

My first comment still holds true. More specifically, the first if test is https://github.com/solarmaterial/amrex/blob/e11cb70c6af7b2d4f97dfc1c357b5eefaf18b87d/test/main.cpp#L86. If that is false, the code is equivalent to
```
int i0;
if (false) {
   i0 = i; // this line will not be run
}
if (i0 < ....) // i0 is uninitialized
```

**Reply by solarmaterial** [reply]
Thanks for all your replies! I'll definitely consider your suggestions. Thanks again for your kindness!

**Reply by solarmaterial** [reply]
Sorry to bother you again, but I still have some questions. I carefully checked the condition of the first `if` test according to your suggestion and used a new method to confirm whether the test was true. I have uploaded the updated code to GitHub.https://github.com/solarmaterial/amrex/tree/main/test

I set the values of `phi_old` to change only with `x` (i.e., `i`), and the output value is set to the integer value of the index:
`phi_old_array(i,j,k,0) = i;`
In this way, the 1D slice output of `phi_old` along the x-direction should be `0, 1, 2, 3, ... 31`.

Now, I give the condition of the `if` statement:
```
if (7 < phi_old_array(i,j,k,0) && phi_old_array(i,j,k,0) < 9)
{
    rhs_array(i,j,k) = 9999; // i = 8
}
```
It is clear to see that the condition is met when `phi_old_array(i,j,k,0) = 8`, and at this time `rhs_array(i,j,k) = 9999`.
I confirm whether the condition is met by checking the 1D slice output of `rhs` along the x-direction. Since the value at the position with index `8` is `9999`, the condition is successfully met.

My question still lies in the relationship between `max_grid_size` and `n_cell`. If I set `n_cell = 32` and `max_grid_size = 8`, it is easy to understand that the computational domain is divided into 4 boxes: `0-7, 8-15, 16-23, 24-31`.

At this time, the node we set earlier is exactly at index `8`, which is the lower boundary of the second box `8-15`. I give the additional code :
```
if (7 < phi_old_array(i,j,k,0) && phi_old_array(i,j,k,0) < 9)
{
    rhs_array(i,j,k) = 9999; // i = 8
    rhs_array(7,j,k) = 8888; // i-1 = 7
    rhs_array(i+1,j,k) = 10000; // i+1 = 9
    rhs_array(i-2,j,k) = 7777; // i-2 = 6
    rhs_array(i+2,j,k) = 11000; // i+2 = 10
}
```
Only the code for `i` greater than or equal to `8` is successfully executed, and the output result is:
```
0     0.015625     0
1     0.046875     0
...
7     0.234375     0
8     0.265625     9999
9     0.296875     10000
10    0.328125     11000
11    0.359375     0
...
31    0.984375     0
```
Since `9999` is present in the output, the test `if (7 < phi_old_array(i,j,k,0) && phi_old_array(i,j,k,0) < 9)` should have been successfully evaluated. Therefore, I guess the reason why the code for `i` less than `8` was not executed can only be that they are in different boxes. 
(
If the index `i` is greater than `15`, and it falls into the third or fourth box, the code remains unexecuted)

Do you have any suggestions for this issue? Thank you very much for your help!

**Comment by WeiqunZhang** [comment]
https://github.com/AMReX-Codes/amrex/discussions/3970#discussioncomment-9656128

**Reply by solarmaterial** [reply]
> Since `9999` is present in the output, the test `if (7 < phi_old_array(i,j,k,0) && phi_old_array(i,j,k,0) < 9)` should have been successfully evaluated.

The first if test should have been true. Did I miss something?

**Comment by WeiqunZhang** [comment]
Again please read
https://github.com/AMReX-Codes/amrex/discussions/3970#discussioncomment-9656128

You asked for a suggestion. My suggestion was there in that link.

**Comment by WeiqunZhang** [comment, marked as answer]
This was what I wrote " How many ghost cells does rhs have? From the
various pieces you have shown, it probably needs 5. And after you fill rhs,
you will need to call SumBounary so that the ghost cell values can be added
to valid cells in other boxes. The special point might be near the box
boundary such that i+1, or i+2, ... or i+5 is in the ghost cell region."

More concisely, you need (1) ghost cells (2) SumBoundary call.

**Reply by solarmaterial** [reply]
I carefully went through your reviews again and, yes, the issue was with the ghost cells. After I set enough ghost cells to the Multifab and called SumBoundary, the prints of the `rhs` was as I expected. Thank you for your patience! I truly appreciate that!
